### PR TITLE
Changes following a discussion with Tommy

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
* Observation
  * We moved out all the getters - this API will provide a nice way to set things ON THE CONTEXT
  * We've added additionalLowCardinality and high cardinality tags on the context, tagsprovider is immutable
  * We removed timing information - there's no gain in unifying the time (e.g. same time for metrics & spans). It's up to the handlers to take control of doing measurements. If a handler is buggy we will see that in its timing information.
  * We removed the throwable getter to explicitly pass the throwable to the handler - that's the only parameter that will never 
change (onError - you have to have an error there)
* Observation.Context
  * It got the displayName and other elements because often they are resolved from the context
  * All the tags are now in the context, we have them in a single spot
* SimpleObservation
  * OnStop will call stop in reverse order (e.g. we call start on handler 1,2,3 so we will call stop in order 3, 2, 1)
* ObservationRegistry
  * We've added `ObservationRegistry` that aggregates all observability related methods
  * We have a dedicated implementation that will automatically register a timer recording handler
  * Added example of a `BiPredicate` for name and context to decide upfront whether the `Observation` should be noop or not
* Fixed some tests
* Added some Javadocs

### CODE REVIEW

File: micrometer-api/src/main/java/io/micrometer/api/instrument/observation/SimpleObservation.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @SuppressWarnings("unchecked")
+    private void notifyOnScopeOpened() {
+        this.handlers.forEach(handler -> handler.onScopeOpened(this, this.context));
+    }
+
+    @SuppressWarnings("unchecked")
+    private void notifyOnScopeClosed() {
+        this.handlers.forEach(handler -> handler.onScopeClosed(this, this.context));
```
Comment (marcingrzejszczak) [inline micrometer-api/src/main/java/io/micrometer/api/instrument/observation/SimpleObservation.java:98]: I think we should do the same reverse iteration here

File: micrometer-api/src/main/java/io/micrometer/api/instrument/observation/ObservationRegistry.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    default Observation start(String name, Observation.Context context) {
+        return observation(name, context).start();
+    }
+
+    Config config();
+
+    /**
+     * Access to configuration options for this registry.
+     */
+    class Config {
```
Comment (marcingrzejszczak) [inline micrometer-api/src/main/java/io/micrometer/api/instrument/observation/ObservationRegistry.java:88]: Add configuration option for asserting whether observation was started when stop, etc. Choose a strategy (throw exception, log error, warning)

File: micrometer-api/src/main/java/io/micrometer/api/instrument/observation/TimerObservationHandler.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    public void onError(Observation observation, Observation.Context context, Throwable throwable) {
+        context.put(Throwable.class, throwable);
+    }
+
+    @Override
+    public void onStop(Observation observation, Observation.Context context) {
+        Timer.Sample sample = context.get(Timer.Sample.class);
+        Tags tags = context.getLowCardinalityTags().and(context.getAdditionalLowCardinalityTags());
+        if (context.containsKey(Throwable.class)) {
+            tags = tags.and(Tag.of("error", context.get(Throwable.class).getLocalizedMessage()));
```
Comment (marcingrzejszczak) [inline micrometer-api/src/main/java/io/micrometer/api/instrument/observation/TimerObservationHandler.java:38]: Add the default `error`  tag that is `none`

File: micrometer-api/src/test/java/io/micrometer/api/instrument/observation/AllMatchingCompositeTimerRecordingHandlerTests.java [hunk 4]
```diff
@@ -29,11 +30,13 @@
 
     MatchingHandler matchingHandler2 = new MatchingHandler();
 
-    Timer.Sample sample = Timer.start(new SimpleMeterRegistry());
+    ObservationRegistry registry = new SimpleObservationRegistry();
+
+    Observation sample = registry.start("hello");
```
Comment (marcingrzejszczak) [inline micrometer-api/src/test/java/io/micrometer/api/instrument/observation/AllMatchingCompositeTimerRecordingHandlerTests.java:35]: call the `observation` method

File: micrometer-test/src/main/java/io/micrometer/core/tck/ObservationRegistryCompatibilityKit.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.isA;
+import static org.mockito.ArgumentMatchers.same;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+/**
+ * Base class for {@link MeterRegistry} compatibility tests.
```
Comment (marcingrzejszczak) [inline micrometer-test/src/main/java/io/micrometer/core/tck/ObservationRegistryCompatibilityKit.java:47]: It's a `PbservationRegistry` now

File: micrometer-api/src/main/java/io/micrometer/api/instrument/observation/SimpleObservation.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        return "{"
+                + "name=" + this.context.getName() + "(" + this.context.getDisplayName() + ")"
+                + ", error=" + this.error
+                + ", context=" + this.context
+                + '}';
+    }
+
+    @SuppressWarnings("unchecked")
+    private void notifyOnObservationStarted() {
+        this.handlers.forEach(handler -> handler.onStart(this, this.context));
```
Comment (jonatan-ivanov) [inline micrometer-api/src/main/java/io/micrometer/api/instrument/observation/SimpleObservation.java:83]: Why are we passing the observation? Now it's kind of empty, what would the handler use it for?

File: micrometer-api/src/main/java/io/micrometer/api/instrument/observation/TimerObservationHandler.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Override
+    public void onStart(Observation observation, Observation.Context context) {
+        Timer.Sample sample = Timer.start(meterRegistry);
+        context.put(Timer.Sample.class, sample);
+    }
+
+    @Override
+    public void onError(Observation observation, Observation.Context context, Throwable throwable) {
+        context.put(Throwable.class, throwable);
```
Comment (jonatan-ivanov) [inline micrometer-api/src/main/java/io/micrometer/api/instrument/observation/TimerObservationHandler.java:30]: If we could set an error field on the context in `Observation::error` we don't need to do this.

**Pull request merged**